content/blog/2025-04-17-blog.md

```diff
@@ -0,0 +1,57 @@
+
+This paper’s monomorphic inline cache cut the cost of a send to a single indirect jump. A few years later, the same group extended the idea to polymorphic inline caches, now standard in V8, HotSpot, PyPy, and Graal. The core intuition—let the first few executions specialize the call-site—remains a workhorse optimization.  
+
+Other ideas that are still used today include: speculative compilation, as profile-guided inlining with cheap de-optimizations fuels every tiered JIT pipeline today; code invalidation via dependency lists, which enable the edit-refresh cycles developers now take for granted; and object shape sharing, which is critical for memory as programs continue to get larger (for example: Chrome routinely juggles millions of JS objects; sharing one descriptor per shape tames this footprint).
+
+However, that is not to say that modern JIT compilers are perfect. Indeed, many of the downsides from SELF's implementation continue to be an issue today. Most notably is the extra startup latency and overhead JIT compilers incure. Customized machine code multiplies binary size, thereby increasing startup time. On mobile, cold-start budgets are measured in tens of milliseconds, prompting projects (such as V8’s Sparkplug baseline) to curb SELF-style mega-methods. Furthermore, the rich PC-to-byte-code maps that make de-optimization and on-the-fly debugging possible also swell metadata. That footprint is incurred even for users who never open DevTools. Finally,  prototype mutation remains a fertile ground for attacks such as prototype-pollution. Static languages dodge whole classes of such issues by construction.
+
+## Looking towards the future
+
+The era of web development was characterized by a constant tug-of-war between static and dynamic compilers. During the 2000s and 2010s, developers embraced the “let the JIT sort it out” mindset and overwhelmingly used JavaScript. Today, the majority of developers are switching to TypeScript, which layers a static type system over top JavaScript. Yet despite this top-level switch in language, at runtime those .ts files are converted to .js files and still funnel into the same hidden-class JS JIT compiler that SELF inspired. The major languages of the 2020s therefore inhabit a middle ground: dynamic under the hood, statically flavored at the surface. That compromise echoes the paper’s own wager that runtime type knowledge, harnessed smartly, can rival compile-time guarantees.
+
+Recently, ARM's big-little cores, WebAssembly’s rise, and the carbon cost of always-on optimisation are reopening the question of whether speculation is worth its wattage. Yet the conceptual toolkit SELF gifted — inline caches, speculative inlining, object shapes, and more — keeps resurfacing. Though today's dynamic languages and JIT compilers look very differeny, the DNA is still recognisably SELF.
```

Comment:
I'm not sure how ARM big.LITTLE fits into this. WebAssembly sure seems like a reaction to the cost of dynamic languages, but I'm not sure I see the connection to heterogeneous multicores?

Reply:
Neat! That's a really interesting take!! If you want to expand on it a little bit, I'm sure it would fit in just great. But I do think it might need a couple of sentences of development to make sense to readers. (Or you could just skip it if you think fully developing the idea wouldn't be worth it!)